xcore/cl_newwavelet_denoise_handler.cpp

```diff
@@ -17,16 +17,367 @@
+        estimate_noise_variance (video_info, buffer->hh[0], buffer->noise_variance);
+        _handler->set_estimated_noise_variation (buffer->noise_variance);
+    } else {
+        _handler->get_estimated_noise_variation (buffer->noise_variance);
+    }
+    return image;
+}
+
+SmartPtr<CLImage>
+CLWaveletNoiseEstimateKernel::get_output_buffer (SmartPtr<DrmBoBuffer> &input, SmartPtr<DrmBoBuffer> &output)
+{
+    XCAM_UNUSED (input);
+    XCAM_UNUSED (output);
+
+    SmartPtr<CLImage> image;
+    SmartPtr<CLWaveletDecompBuffer> buffer = _handler->get_decomp_buffer (_channel, _current_layer);
+
+    if (_subband == CL_WAVELET_SUBBAND_HL) {
+        image = buffer->hl[1];
+    } else if (_subband == CL_WAVELET_SUBBAND_LH) {
+        image = buffer->lh[1];
+    } else if (_subband == CL_WAVELET_SUBBAND_HH) {
+        image = buffer->hh[1];
+    } else {
+        image = buffer->ll;
+    }
+    return image;
+}
+
+XCamReturn
+CLWaveletNoiseEstimateKernel::prepare_arguments (
+    SmartPtr<DrmBoBuffer> &input, SmartPtr<DrmBoBuffer> &output,
+    CLArgument args[], uint32_t &arg_count,
+    CLWorkSize &work_size)
+{
+    SmartPtr<CLContext> context = get_context ();
+    const VideoBufferInfo & video_info_out = output->get_video_info ();
+
+    _image_in = get_input_buffer (input, output);
+    _image_out = get_output_buffer (input, output);
+
+    XCAM_ASSERT (_image_in->is_valid () && _image_out->is_valid ());
+    XCAM_FAIL_RETURN (
+        WARNING,
+        _image_in->is_valid () && _image_out->is_valid (),
+        XCAM_RETURN_ERROR_MEM,
+        "cl image kernel(%s) in/out memory not available", get_kernel_name ());
+
+    //set args;
+    args[0].arg_adress = &_image_in->get_mem_id ();
+    args[0].arg_size = sizeof (cl_mem);
+    args[1].arg_adress = &_image_out->get_mem_id ();
+    args[1].arg_size = sizeof (cl_mem);
+    arg_count = 2;
+
+    work_size.dim = XCAM_DEFAULT_IMAGE_DIM;
+    work_size.local[0] = 8;
+    work_size.local[1] = 4;
+    if (_channel == CL_WAVELET_CHANNEL_Y) {
+        work_size.global[0] = XCAM_ALIGN_UP ((video_info_out.width >> _current_layer) / 4, work_size.local[0]);
+        work_size.global[1] = XCAM_ALIGN_UP ((video_info_out.height >> _current_layer) / 4, work_size.local[1]);
+    } else if (_channel == CL_WAVELET_CHANNEL_UV) {
+        work_size.global[0] = XCAM_ALIGN_UP ((video_info_out.width >> _current_layer) / 4, work_size.local[0]);
+        work_size.global[1] = XCAM_ALIGN_UP ((video_info_out.height >> (_current_layer + 1)) / 4, work_size.local[1]);
+    }
+
+    return XCAM_RETURN_NO_ERROR;
+}
+
+XCamReturn
+CLWaveletNoiseEstimateKernel::estimate_noise_variance (const VideoBufferInfo & video_info, SmartPtr<CLImage> image, float* noise_var)
+{
+    XCamReturn ret = XCAM_RETURN_NO_ERROR;
+
+    SmartPtr<CLEvent> map_event = new CLEvent;
+    void *buf_ptr = NULL;
+
+    CLImageDesc cl_desc;
+    CLImage::video_info_2_cl_image_desc (video_info, cl_desc);
```

Comment:
video info is the original input buffer's info. if only want to get the width/heihgt/format, maybe
cl_desc = image->get_image_desc () is a better choice.